tcmur_cmd_handler.c

```diff
@@ -216,35 +218,22 @@ static struct unmap_state *unmap_state_alloc(struct tcmu_device *dev,
 	return NULL;
 }
 
-static void unmap_state_free(struct unmap_state *state)
+static void unmap_state_put(struct tcmu_device *dev,
+			    struct tcmulib_cmd *origcmd, int ret)
 {
-	pthread_mutex_destroy(&state->lock);
-	free(state);
-}
-
-static void handle_unmap_cbk(struct tcmu_device *dev, struct tcmulib_cmd *ucmd,
-			     int ret)
-{
-	struct unmap_descriptor *desc = ucmd->cmdstate;
-	struct tcmulib_cmd *origcmd = desc->origcmd;
 	struct unmap_state *state = origcmd->cmdstate;
 	bool error;
 	int status;
 
-	free(desc);
-
 	pthread_mutex_lock(&state->lock);
 	error = state->error;
 	/*
 	 * Make sure to only copy the first error
 	 */
 	if (!error && ret) {
 		state->error = true;
-		state->status = ret;
```

Comment:
I do not think this will work.

If you have sent 2 unmap calls to handle the command, and if the first one fails then we will set state->error = true, but not set state->status. When the second call completes, it might succeed. Refcount will now be zero and aio_command_finish will see

error = true
status = for  TCMU_STS_ASYNC_HANDLED if we are called from handle_unmap_in_writesame and TCMU_STS_OK if called from handle_unmap->handle_unmap_internal.
ret = TCMU_STS_OK

so because error=true from the first call, TCMU_STS_ASYNC_HANDLED or TCMU_STS_OK will be returned by aio_command_finish.

Reply:
I see that now, thanks!